### PR TITLE
[On Hold] Batch cache fetch

### PR DESCRIPTION
**Pull request by adgad**
This PR maintains the benefit of batched requests to CAPI whilst still caching individual articles.

Number of fetches on a clean startup reduced from ~114 to ~22.

TODO:
- [ ] Add some comments to the code, and maybe tidy up a little
- [ ] Compare against production results

### CODE REVIEW

File: server/lib/cache.js [hunk 1]
```diff
@@ -51,33 +51,93 @@ class Cache {
...
+
+			eventualData =
+				this._fetch(listKey, itemsToFetch, now, ttl, batchFetcher)
+				.then((items) => {
+					if(items) {
+						items.forEach((item, index) => {
+							promises.set(itemsToWaitFor[index], item);
+						});
+					}
+					return [...promises.values()];
```
Comment (charypar) [inline server/lib/cache.js]: You don't wan to wait for the fetch to finish here, do you? You're returning stale, these are `itemsToRefresh` basically.

**Comment by charypar** [comment]
@adgad I think really stupidly OCD naming of the variables in the caching logic will help figure out what it should do, it's still pretty confusing at the moment :)

File: server/lib/cache.js [hunk 2]
```diff
@@ -51,33 +51,93 @@ class Cache {
 		}
 	}
 
-	_fetch(key, now, ttl, fetcher) {
-		const metricsKey = key.split('.')[0];
+	// Caching for lists. For a given list of IDs, cache the IDs individually, but fetch in a batch
+	// when cache expires return stale data immediately and fetches fresh one
+	cachedList(listKey, ids, ttl, batchFetcher) {
```
Comment (charypar) [inline server/lib/cache.js]: I'd suggest calling the things here `results` which are promises you'll return, `itemsToWaitFor` and `itemsToRefresh`.

- `itemsToWaitFor` are the ones we don't even have stale, so those go into `results`
- `itemsToRefresh` we have stale, so they only kick of a fetch that updates cache when done
- `results` are a mixture of resolved and sometimes unresolved promises for all the stuff

File: server/lib/cache.js [hunk 3]
```diff
@@ -51,33 +51,93 @@ class Cache {
...
+	}
 
-			this.contentCache[key] = {
-				expire: expireTime,
-				data: it
-			};
+	_fetch(key, ids, now, ttl, fetcher) {
+		const requestKey = ids && ids.length ? `${key}.${ids.join('_')}` : key;
+		const metricsKey = key.split('.')[0];
+		const expireTime = now + ttl;
```
Comment (charypar) [inline server/lib/cache.js]: You probably want to compute that when the item arrives, not when you kick of the fetch?

File: server/lib/cache.js [hunk 4]
```diff
@@ -51,33 +51,93 @@ class Cache {
...
+		if(this.requestMap[requestKey])
+			return this.requestMap[requestKey];
 
-			delete this.requestMap[key];
+		metrics.count(`cacher.${metricsKey}.fresh`, 1);
+
+		this.requestMap[requestKey] = fetcher(ids)
+		.then((it) => {
+			if(ids && ids.length) {
```
Comment (charypar) [inline server/lib/cache.js]: This feels like both branches could be unified by extracting `id` from `key` when only one thing is being fetched and using the `if` branch every time.

File: server/lib/cache.js (hunk 2 above)
Comment (charypar) [inline server/lib/cache.js]: And then you probably want to return `Promise.all(promises)` in all the cases (might mean you need to be able to `.catch` on each of the remote ones to always return as much as you can).

**Comment by adgad** [comment]
Putting this on hold for now - it's a lot of added code complexity for potentially little gain (especially after https://github.com/Financial-Times/next-graphql-api/issues/26)

**Comment by charypar** [comment]
I agree, we'll still need caching even after #26 is done, but runaway cache doesn't seem to be the cause of the leak anyway.

**Comment by adgad** [comment]
Idea to simplify this: give lists of uuid's a cache key (which would be the list they originated from)

something like: `backend.contentv1(['a','b','c'], 'list_<list-uuid>')` or `backend.contentv1(['a','b','c'], 'relatedContent_<content-uuid>')`

**Pull request closed without merging**